### ISSUE TITLE
re: selectors

### ISSUE THREAD

**Issue by faceyspacey** [dai-shi/reactive-react-redux#9]
@dai-shi @theKashey @ScriptedAlchemy

So I've been doing some thinking around selectors, and I've come to the conclusion that the babel plugin approach is a horrible path to hell. I'll describe why in a second. Here's the alternative:

```js
function MyComponent(props) {
   const state = useReduxState(props, someArgEtc) // here's the important line
   return <div>{state.fooSelector(props, someArgEtc)}</div>
}
```

===

```js
function MyComponent(props) {
   const state = useReduxState()
   return useMemo(<div>{state.fooSelector(props, someArgEtc)}</div>, [props, someArgEtc])
}
```

So, basically we make the user/developer do a little bit of work by providing the dependencies to track. And we dont make em think about `useMemo`. 

---
Of course to do this properly, rests on `useReduxState` knowing about selectors, so therefore Redux must be forked to have this `createStore` API:

```js
const selectors = {
   fooSelector: (state, props, someArgEtc) => ...,
   anotherSelector: (state, etc) => ...
}
const store = createStore(reducers, selectors) // both are automatically "combined"
```

Now, we know what selectors we have like reducers. Therefore, the proxy can  prevent re-render from within the hook. It does so by shallow comparison of the argument dependencies.

Hell, we could even accomplish this without a fork of redux. `useReduxState` can accept additional args that are simply compared for changes. *And of course it's up to the developer to match with actual usage of selectors.*

----

### Why we dont wanna mess with babel?

Here's why:

```js
function RespondComponent(props, state, actions) {
    if (props.foo) return <div>{state.selectorA(props)}</div>
    return <div>{state.selectorB(props)}</div>
}
``` 

-->

```js
function RespondComponent(props, state, actions) {
    if (props.foo) {
        const valA = state.selectorA(props)
        return useMemo(<div>{valA}</div>, [valA])
    }

     const valB = state.selectorB(props)
     return useMemo(<div>{valB}</div>, [valB])
}
``` 

And that's the easiest possible example!

If making all the appropriate visitations of selectors in search of selectors is even 100% reliable, then taking this to the next level where **selectors are correctly conditionally depended upon** is even more error-prone.

Also, nobody talks about babel compilation speed, but this stuff adds up. Nobody wants compilation time decreased because of this stuff. 

Anyway the **reliability factor** is just too low based on my experience doing babel plugins. 

----

Now that all said, I think we've found the perfect solution for how to handle this in runtime code *within existing proxies*! The only caveat is you have to have a store creation mechanism like in *Respond Framework*. Which isn't really a caveat since that's the exact direction we're going, and since we could also just give `useReduxState` additional super powers of bail-out capabilities through passing it deps, as if it was `useMemo`. 

Any tips to make this a reality are much appreciated.

**Comment by dai-shi** [comment]
I don't think I understand everything around the purpose of selectors, but if you want to pass `deps`, I'd do something like this:

```javascript
const triggerReducer = (s, a) => a || !s;
const useTriggerUpdate = () => useReducer(triggerReducer, false)[1];

export const useReduxState = (deps) => {
  const lastDeps = useRef(null);
  useEffect(() => {
    lastDeps.current = deps;
  }, deps);
  const triggerUpdate = useTriggerUpdate();
  // ...snipped...
        triggerUpdate(lastDeps.current); // instead of forceUpdate()
  // ...snipped...
};
```

Note that I haven't tested and I am not 100% confident.

**Comment by theKashey** [comment]
1. If you will be able to get `selectors` from a some _centralized_ `store` - we might track their usage as well as state usage. Plus we might track state usage within `selector`. 👍 
2. We should be able to detect usages outside of selectors, and ask user not to do it. 
3. @dai-shi  - selectors are a small helper functions, which does "the same" work every time. It's more _predictable_ to use them, than _randomly_ read from a state. They also enable deep memozation, and cuts all updates of a state, which were important, but didn't lead to a `component state` update. Example - todo list, showing only tasks assigned to me, and you are adding a task for yourself - the state is updated, todo list is updated, but "my tasks" are not.

**Comment by theKashey** [comment]
From another point of view - the initial idea was to throw aways `mapStateToProps`, AND `selectors`. Isn't it?

**Comment by faceyspacey** [comment]
> If you will be able to get selectors from a some centralized store - we might track their usage as well as state usage. Plus we might track state usage within selector. 👍

Mo%$erFu$&in excellent!!

Funny thing, is i went to bed last night realizing the same thing--that the selector problem is no different than reducers once they come "from some centralized" store as you mentioned, i.e:

```js
createStore(reducer, selectors)
```

Which is exactly the plan.

> From another point of view - the initial idea was to throw aways mapStateToProps, AND selectors. Isn't it?

Throw away `mapStateToProps`, but not selectors. But yes, move them away from having to be in every component within `mapStateToProps`. And instead define them once in the "centralized" store *with smart caching for use in potentially multiple components*. 

This means **"memoization"** is no longer what we want, but rather **caching of multiple values**, which is why I was looking deeply into your intriguing [kashe](https://github.com/theKashey/kashe) library which seems to be a best of both worlds approach--those approaches being:

- cache multiple values 
- prevent memory leaks via the automatic garbage collection capability of weakmaps

 It seems, that this capability exists *even* without the added hierarchy provided by the `boxed` and `inboxed` additional capabilities the library provides. Am I correct?

Basically, am I seeing a way to move past the *reselect* and *rereselect* problems where a selector function must be duplicated or created from a factory for every component that uses it. And therefore we can just define it once on store creation, as if we are modeling a `UI Database`, more or less like MobX + MobX State Tree, but the more *pure* Redux version. 

### Please inform me if I'm seeing clearly how to use your tools and if my plan is compatible with your great wisdom

**Comment by dai-shi** [comment]
@theKashey 

> selectors are a small helper functions, which does "the same" work every time.

I understand that in general.

> It's more predictable to use them, than randomly read from a state.

What does "randomly read" mean?

> They also enable deep memozation

Can you simply use `useMemo`?

**Comment by faceyspacey** [comment]
`useMemo` can be used in userland. but as above, to handle all scenarios with babel, and automatically do this for the user is not reliable. at the very least, it's extremely challenging to achieve 100% reliability. 

dai-shi, i'm gonna be releasing docs for Respond Framework soon. I'm putting together an informal team of allstars. I want to extend you an invitation to join us. There will be 4-5 core team members. You won't have to do anymore than you want to. It could be just a badge of honor if you want. Basically I want to give credit to the fundamental work you've done here, as we do a lot to promote a beautiful new API for the Redux ecosystem.

**Comment by theKashey** [comment]
> Please inform me if I'm seeing clearly how to use your tools and if my plan is compatible with your great wisdom

TLDR - @faceyspacey - you get it right.

> What does "randomly read" mean?

A wild some, where you could discover everything. See below.

> Can you simply use useMemo?

Divide and conquer. Use memo and "random reads" are one BIG block, which got executed as a whole. You cannot execute a part.

- memoization is not working for a big thing - some _fluctuations_ are killing it
- it's imposible to predict how some value change would affect a "result", if that one thing got used _twice_.
- `proxyequal` is relying on _selector usage_, ie - if you are looking _inside_ an object - you need value inside, not the object. Which could be false for a _big_ things.

But splitting A THING into smaller parts, and assembling them back - you are making everything more stable, and easy to reason about.

PS: ^that^ is not a thing, if you have many small and simple components, but the first (even a single one) big one, would kill all the fun.

**Comment by theKashey** [comment]
> And Anton, as you know, the invitation has been extended since basically last year. Zack and I already consider you part of our team.

invite me to the organization then :trollface:

**Comment by faceyspacey** [comment]
The organization is called:

### Devotion App Development

picture this:

![album cover](https://images-na.ssl-images-amazon.com/images/I/51TUNA8ww1L.jpg)

On top, instead of ***Guns 'N' Roses*** it says:

*Respond Framework*

on the bottom, it says:

*Devotion 'App' Development*


So we will all have skull cartoons made in Illustrator. And more importantly, we aren't an "application development team"--we are a **Band**, and *Respond Framework* is our first album!


☠D'A'D☠

Here's our other big symbol:

🙏

**Comment by faceyspacey** [comment]
<a href="http://www.youtube.com/watch?feature=player_embedded&v=iRKxQ-8pq-M#t=6m0s
" target="_blank"><img src="http://img.youtube.com/vi/iRKxQ-8pq-M/0.jpg" 
alt="IMAGE ALT TEXT HERE" width="240" height="180" border="10" /></a>

I'm gonna use CrazyTalk Animator (or Adobe Character Animator) to animate a cartoon of my/our skull for video tutorials. And then we have a screencast of code, and then in the lower right corner is a talking skull. It should be a piece of cake with these modern tools.

**Comment by faceyspacey** [comment]
# ps. we operate beyond all societal constructs. we dont need a taxable corporation, github organization, or anything, but to exist in our ***own minds*** and the minds of our fans

aint that right @ScriptedAlchemy !

**Comment by ScriptedAlchemy** [comment]
Late to the party here! For context, I work with @faceyspacey :) 

@dai-shi @theKashey you've done some fantastic work on this project. I've become a heavy promoter and dependent on this project, and I intend to roll react-hooks-easy-redux out to all major companies I work for. This is a substantial improvement in DX compared to react-redux!

Id love to work with you guys in whatever capacity you may have. OSS is a labor of love and I know time is always working against us all - but just this thread and the work to come from is has been really awesome - hard problems solved! It would be fun for us to work together especially as we all seem geared to tackling the exciting challenges that are a little harder to solve. We are all busy, but some effective collaboration could user in a profound new era of React development.

**Comment by dai-shi** [comment]
Closing this in favor of #16 and #19. Please file a new issue for further discussion.
